File: src/main/java/chapter4/etc/MathFunctionsExample.java
package chapter4.etc;

import common.Log;
import hu.akarnokd.rxjava2.math.MathFlowable;
import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;


public class MathFunctionsExample {

    public void marbleDiagram() {
        Integer[] data = {1, 2, 3, 4};

        // 1. Count
        Single<Long> source = Observable.fromArray(data)
                .count();
        source.subscribe(count -> Log.it("count is " + count));

        // 2. max() & min()
        Flowable.fromArray(data)
                .to(MathFlowable::max)
                .subscribe(max -> Log.it("max is " + max));

        Flowable.fromArray(data)
                .to(MathFlowable::min)
                .subscribe(min -> Log.it("min is " + min));

        // 3. sum() & average
        Flowable<Integer> flowable = Flowable.fromArray(data)
                .to(MathFlowable::sumInt);
        flowable.subscribe(sum -> Log.it("sum is " + sum));

        Flowable<Double> flowable2 = Observable.fromArray(data)
                .toFlowable(BackpressureStrategy.BUFFER)
                .to(MathFlowable::averageDouble);
        flowable2.subscribe(avg -> Log.it("average is " + avg));
    }

    public static void main(String[] args) {
        MathFunctionsExample demo = new MathFunctionsExample();
        demo.marbleDiagram();
    }

}

File: src/main/kotlin/chapter4/etc/MathFunctionsExample_.kt
package chapter4.etc

import common.Log_
import hu.akarnokd.rxjava2.math.MathFlowable
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.Observable


class MathFunctionsExample_ {

    fun marbleDiagram() {
        val data: Array<Int> = arrayOf(1, 2, 3, 4)

        // 1. Count
        val source = Observable.fromArray(*data)
                .count()
        source.subscribe { count -> Log_.it("count is $count") }

        // 2. max() & min()
        Flowable.fromArray(*data)
                .to { MathFlowable.max(it) }
        source.subscribe { max -> Log_.it("max is $max") }

        Flowable.fromArray(*data)
                .to { MathFlowable.min(it) }
                .subscribe { min -> Log_.it("min is $min") }

        // 3. sum() & average
        val flowable = Flowable.fromArray(*data)
                .to { MathFlowable.sumInt(it) }
        flowable.subscribe { sum -> Log_.it("sum is $sum") }

        val flowable2 = Observable.fromArray(*data)
                .toFlowable(BackpressureStrategy.BUFFER)
                .to { MathFlowable.averageDouble(it) }
        flowable2.subscribe { avg -> Log_.it("average is $avg") }
    }

}


fun main(args: Array<String>) {
    val demo = MathFunctionsExample_()
    demo.marbleDiagram()
}
